diff --git a/Makefile.am b/Makefile.am
index 2638e52f..392c8bdc 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -108,6 +108,7 @@ dyndev_SRC =  \
   cardrdr.c   \
   comm3705.c  \
   commadpt.c  \
+  cxip.c      \
   tcpnje.c    \
   con1052c.c  \
   console.c   \
@@ -187,6 +188,7 @@ HERCLIBS2 =     \
   libhdt3420_not_mod.la
 
 HERCMODS =      \
+  cxip.la       \
   dyncrypt.la   \
   dyngui.la     \
   hdteq.la      \
@@ -270,6 +272,10 @@ EXTRA_PROGRAMS = hercifc
 #   ModuleName_la_DEPENDENCIES = libherc.la (may not be necessary)
 #----------------------------------------------------------------------------
 
+cxip_la_SOURCES     = cxip.c
+cxip_la_LDFLAGS     = $(DYNMOD_LD_FLAGS)
+cxip_la_LIBADD      = $(DYNMOD_LD_ADD)
+
 dyncrypt_la_SOURCES = dyncrypt.c
 dyncrypt_la_LDFLAGS = $(DYNMOD_LD_FLAGS)
 dyncrypt_la_LIBADD  = $(DYNMOD_LD_ADD)
diff --git a/Makefile.in b/Makefile.in
index 24745886..11e0e556 100644
--- a/Makefile.in
+++ b/Makefile.in
@@ -171,13 +171,19 @@ LTLIBRARIES = $(lib_LTLIBRARIES) $(modexec_LTLIBRARIES) \
 am__DEPENDENCIES_1 =
 am__DEPENDENCIES_2 = libherc.la libhercs.la libhercu.la \
 	$(am__DEPENDENCIES_1) $(am__DEPENDENCIES_1)
-dyncrypt_la_DEPENDENCIES = $(am__DEPENDENCIES_2)
-am_dyncrypt_la_OBJECTS = dyncrypt.lo
-dyncrypt_la_OBJECTS = $(am_dyncrypt_la_OBJECTS)
+cxip_la_DEPENDENCIES = $(am__DEPENDENCIES_2)
+am_cxip_la_OBJECTS = cxip.lo cxip_protocol.lo
+cxip_la_OBJECTS = $(am_cxip_la_OBJECTS)
 AM_V_lt = $(am__v_lt_@AM_V@)
 am__v_lt_ = $(am__v_lt_@AM_DEFAULT_V@)
 am__v_lt_0 = --silent
 am__v_lt_1 = 
+cxip_la_LINK = $(LIBTOOL) $(AM_V_lt) --tag=CC $(AM_LIBTOOLFLAGS) \
+	$(LIBTOOLFLAGS) --mode=link $(CCLD) $(AM_CFLAGS) $(CFLAGS) \
+	$(cxip_la_LDFLAGS) $(LDFLAGS) -o $@
+dyncrypt_la_DEPENDENCIES = $(am__DEPENDENCIES_2)
+am_dyncrypt_la_OBJECTS = dyncrypt.lo
+dyncrypt_la_OBJECTS = $(am_dyncrypt_la_OBJECTS)
 dyncrypt_la_LINK = $(LIBTOOL) $(AM_V_lt) --tag=CC $(AM_LIBTOOLFLAGS) \
 	$(LIBTOOLFLAGS) --mode=link $(CCLD) $(AM_CFLAGS) $(CFLAGS) \
 	$(dyncrypt_la_LDFLAGS) $(LDFLAGS) -o $@
@@ -648,7 +654,8 @@ am__depfiles_remade = ./$(DEPDIR)/_archdep_templ.Plo \
 	./$(DEPDIR)/convto64.Po ./$(DEPDIR)/cpu.Plo \
 	./$(DEPDIR)/crypto.Plo ./$(DEPDIR)/ctc_ctci.Plo \
 	./$(DEPDIR)/ctc_lcs.Plo ./$(DEPDIR)/ctc_ptp.Plo \
-	./$(DEPDIR)/ctcadpt.Plo ./$(DEPDIR)/dasdcat.Po \
+	./$(DEPDIR)/ctcadpt.Plo ./$(DEPDIR)/cxip.Plo \
+	./$(DEPDIR)/cxip_protocol.Plo ./$(DEPDIR)/dasdcat.Po \
 	./$(DEPDIR)/dasdconv.Po ./$(DEPDIR)/dasdconv64.Po \
 	./$(DEPDIR)/dasdcopy.Po ./$(DEPDIR)/dasdcopy64.Po \
 	./$(DEPDIR)/dasdinit.Po ./$(DEPDIR)/dasdinit64.Po \
@@ -737,14 +744,14 @@ AM_V_CCLD = $(am__v_CCLD_@AM_V@)
 am__v_CCLD_ = $(am__v_CCLD_@AM_DEFAULT_V@)
 am__v_CCLD_0 = @echo "  CCLD    " $@;
 am__v_CCLD_1 = 
-SOURCES = $(dyncrypt_la_SOURCES) $(dyngui_la_SOURCES) \
-	$(hdt1052c_la_SOURCES) $(hdt1403_la_SOURCES) \
-	$(hdt2703_la_SOURCES) $(hdt2880_la_SOURCES) \
-	$(hdt3088_la_SOURCES) $(hdt3270_la_SOURCES) \
-	$(hdt3420_la_SOURCES) $(hdt3505_la_SOURCES) \
-	$(hdt3525_la_SOURCES) $(hdt3705_la_SOURCES) \
-	$(hdtdummy_la_SOURCES) $(hdteq_la_SOURCES) \
-	$(hdtptp_la_SOURCES) $(hdtqeth_la_SOURCES) \
+SOURCES = $(cxip_la_SOURCES) $(dyncrypt_la_SOURCES) \
+	$(dyngui_la_SOURCES) $(hdt1052c_la_SOURCES) \
+	$(hdt1403_la_SOURCES) $(hdt2703_la_SOURCES) \
+	$(hdt2880_la_SOURCES) $(hdt3088_la_SOURCES) \
+	$(hdt3270_la_SOURCES) $(hdt3420_la_SOURCES) \
+	$(hdt3505_la_SOURCES) $(hdt3525_la_SOURCES) \
+	$(hdt3705_la_SOURCES) $(hdtdummy_la_SOURCES) \
+	$(hdteq_la_SOURCES) $(hdtptp_la_SOURCES) $(hdtqeth_la_SOURCES) \
 	$(hdttcph_la_SOURCES) $(hdttcpnje_la_SOURCES) \
 	$(hdtzfcp_la_SOURCES) $(libhdt3420_not_mod_la_SOURCES) \
 	$(libherc_la_SOURCES) $(EXTRA_libherc_la_SOURCES) \
@@ -764,14 +771,14 @@ SOURCES = $(dyncrypt_la_SOURCES) $(dyngui_la_SOURCES) \
 	$(hetupd_SOURCES) $(maketape_SOURCES) $(tapecopy_SOURCES) \
 	$(tapemap_SOURCES) $(tapesplt_SOURCES) $(tfprint_SOURCES) \
 	$(tfswap_SOURCES) $(txt2card_SOURCES) $(vmfplc2_SOURCES)
-DIST_SOURCES = $(dyncrypt_la_SOURCES) $(dyngui_la_SOURCES) \
-	$(hdt1052c_la_SOURCES) $(hdt1403_la_SOURCES) \
-	$(hdt2703_la_SOURCES) $(hdt2880_la_SOURCES) \
-	$(hdt3088_la_SOURCES) $(hdt3270_la_SOURCES) \
-	$(hdt3420_la_SOURCES) $(hdt3505_la_SOURCES) \
-	$(hdt3525_la_SOURCES) $(hdt3705_la_SOURCES) \
-	$(hdtdummy_la_SOURCES) $(hdteq_la_SOURCES) \
-	$(hdtptp_la_SOURCES) $(hdtqeth_la_SOURCES) \
+DIST_SOURCES = $(cxip_la_SOURCES) $(dyncrypt_la_SOURCES) \
+	$(dyngui_la_SOURCES) $(hdt1052c_la_SOURCES) \
+	$(hdt1403_la_SOURCES) $(hdt2703_la_SOURCES) \
+	$(hdt2880_la_SOURCES) $(hdt3088_la_SOURCES) \
+	$(hdt3270_la_SOURCES) $(hdt3420_la_SOURCES) \
+	$(hdt3505_la_SOURCES) $(hdt3525_la_SOURCES) \
+	$(hdt3705_la_SOURCES) $(hdtdummy_la_SOURCES) \
+	$(hdteq_la_SOURCES) $(hdtptp_la_SOURCES) $(hdtqeth_la_SOURCES) \
 	$(hdttcph_la_SOURCES) $(hdttcpnje_la_SOURCES) \
 	$(hdtzfcp_la_SOURCES) $(libhdt3420_not_mod_la_SOURCES) \
 	$(libherc_la_SOURCES) $(EXTRA_libherc_la_SOURCES) \
@@ -1194,6 +1201,7 @@ HERCLIBS2 = \
   libhdt3420_not_mod.la
 
 HERCMODS = \
+  cxip.la       \
   dyncrypt.la   \
   dyngui.la     \
   hdteq.la      \
@@ -1226,6 +1234,9 @@ lib_LTLIBRARIES = $(HERCLIBS2)
 #   ModuleName_la_LIBADD       = libherc.la (the Core Hercules Shared Library)
 #   ModuleName_la_DEPENDENCIES = libherc.la (may not be necessary)
 #----------------------------------------------------------------------------
+cxip_la_SOURCES = cxip.c cxip_protocol.c
+cxip_la_LDFLAGS = $(DYNMOD_LD_FLAGS)
+cxip_la_LIBADD = $(DYNMOD_LD_ADD)
 dyncrypt_la_SOURCES = dyncrypt.c
 dyncrypt_la_LDFLAGS = $(DYNMOD_LD_FLAGS)
 dyncrypt_la_LIBADD = $(DYNMOD_LD_ADD)
@@ -2208,6 +2219,8 @@ clean-noinstLTLIBRARIES:
 	  echo rm -f $${locs}; \
 	  rm -f $${locs}; \
 	}
+cxip.la: $(cxip_la_OBJECTS) $(cxip_la_DEPENDENCIES) $(EXTRA_cxip_la_DEPENDENCIES) 
+	$(AM_V_CCLD)$(cxip_la_LINK) -rpath $(modexecdir) $(cxip_la_OBJECTS) $(cxip_la_LIBADD) $(LIBS)
 
 dyncrypt.la: $(dyncrypt_la_OBJECTS) $(dyncrypt_la_DEPENDENCIES) $(EXTRA_dyncrypt_la_DEPENDENCIES) 
 	$(AM_V_CCLD)$(dyncrypt_la_LINK) -rpath $(modexecdir) $(dyncrypt_la_OBJECTS) $(dyncrypt_la_LIBADD) $(LIBS)
@@ -2504,6 +2517,8 @@ distclean-compile:
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/ctc_lcs.Plo@am__quote@ # am--include-marker
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/ctc_ptp.Plo@am__quote@ # am--include-marker
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/ctcadpt.Plo@am__quote@ # am--include-marker
+@AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/cxip.Plo@am__quote@ # am--include-marker
+@AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/cxip_protocol.Plo@am__quote@ # am--include-marker
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/dasdcat.Po@am__quote@ # am--include-marker
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/dasdconv.Po@am__quote@ # am--include-marker
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/dasdconv64.Po@am__quote@ # am--include-marker
@@ -3110,6 +3125,8 @@ distclean: distclean-recursive
 	-rm -f ./$(DEPDIR)/ctc_lcs.Plo
 	-rm -f ./$(DEPDIR)/ctc_ptp.Plo
 	-rm -f ./$(DEPDIR)/ctcadpt.Plo
+	-rm -f ./$(DEPDIR)/cxip.Plo
+	-rm -f ./$(DEPDIR)/cxip_protocol.Plo
 	-rm -f ./$(DEPDIR)/dasdcat.Po
 	-rm -f ./$(DEPDIR)/dasdconv.Po
 	-rm -f ./$(DEPDIR)/dasdconv64.Po
@@ -3350,6 +3367,8 @@ maintainer-clean: maintainer-clean-recursive
 	-rm -f ./$(DEPDIR)/ctc_lcs.Plo
 	-rm -f ./$(DEPDIR)/ctc_ptp.Plo
 	-rm -f ./$(DEPDIR)/ctcadpt.Plo
+	-rm -f ./$(DEPDIR)/cxip.Plo
+	-rm -f ./$(DEPDIR)/cxip_protocol.Plo
 	-rm -f ./$(DEPDIR)/dasdcat.Po
 	-rm -f ./$(DEPDIR)/dasdconv.Po
 	-rm -f ./$(DEPDIR)/dasdconv64.Po
